Upgrade ELK stack to v8.x

caffeinatedpixel opened the pull request:
Closes #74

#### BeaKer Server Changes:
- Adds the ability to upgrade from Elastic version 7.5.2 to 7.17.9, and subsequently 8.7.0
	- For upgrade installs, v7.17.9 will be installed and then the installer will end. If a user wants to upgrade to v8, the user will have to run the installer again when they're ready to.
	- Elastic v7.17.9 will support Winlogbeat logs from v7.5.2 - v7.17.9
	- Elastic v8.7.0 will support Winlogbeat logs from v7.17.0 - 8.7.0, but already ingested logs using the `sysmon-*` index will still be visible in Kibana. New logs from <7.17.9 will **not** be ingested.
- Installer now creates the `elastic/kibana` service account token automatically, which is now required for Kibana 
	- Removed the `elasticsearch.username` and `elasticsearch.password` entries in `kibana.yml` as they are deprecated in favor of the service account token
- Installer now creates required encryption keys for Kibana plugins to work properly
- Installer creates a snapshot repository named `beaker` and attempts to create a snapshot before upgrading to 8.7.0
- Installer loads index templates, index lifecycle policy, aliases, data streams, and ingest pipelines for the matching Winlogbeat version
	- For Elastic v7.17.9, logs are sent to a data stream that uses the index pattern `winlogbeat-7.17.9`
	- For Elastic v8.7.0, logs are sent to a data stream that uses the index pattern `winlogbeat-8.7.0` , and is processed by an ingest pipeline named `winlogbeat-8.7.0-routing`
- Updates the Sysmon and Beacon Kibana dashboards to include both the `sysmon-YYYY-MM-dd` (legacy index pattern), and `winlogbeat-%{[agent.version]}` pattern.
- Creates the CA before generating any certificates, as the previous usage is now deprecated. Generates CA and certificates using the root user.
- Adds  `./kibana/check_kibana.py` in order to parse the Kibana status API response and check whether or not Kibana is finished upgrading
	- Not waiting for the saved objects migration will cause the Kibana dashboard upload to fail
- Adds an `ELK_STACK_VERSION` environment variable to make it easier to detect the currently installed Elastic version
- Adds `./elasticsearch/export_day.sh` in order to export a day's worth of data from the `winlogbeat-*` indices (emulates the daily indices from previous versions)

#### BeaKer Agent Changes:

- Changes the install directory from `C:\Program Files\winlogbeat-*` to `C:\Program Files\Winlogbeat-BeaKer` in order to be able to install newer versions of winlogbeat and better detect the currently installed version
- Adds a parameter, `BeatsVersion`, that allows the desired version of winlogbeat to be installed to be overriden
- Updates the `winlogbeat.yml` config as needed for 7.17 and 8.x 
- Detects whether or not an intermediate upgrade to 7.17 is needed

Testing (Ubuntu 20.04):
- Upgraded BeaKer v0.13 install and verified that v7.17 was installed and not v8.7.0. Ensured that the service account token, index template, ILM policy, data stream, encryption keys, and dashboards were all created. Ensured that data from the previous version was still visible in the dashboards. Ensured that data from winlogbeat v7.5.2 was still being ingested. Ensured that data from winlogbeat v7.17.9 was being ingested via the `winlogbeat-7.17.9` data stream.
- Upgraded the same install to v8.7.0. Ensured that the prompt asking whether to upgrade or reinstall the current version was displayed. Ensured that the prompt asking to create a snapshot before upgrade was displayed and the snapshot was created. Ensured that the ingest pipelines were loaded, Kibana's migration status completed, and that the dashboards were loaded. Ensured that data coming from both winlogbeat 7.17.9 and 8.7.0 was being ingested into Elasticsearch.
- Ensured that a fresh install installs v8.7.0 and completes successfully.

Zalgo2462 commented on installer/stage/BeaKer/install_beaker.sh:
```diff
@@ -160,6 +236,352 @@ ensure_certificates_exist () {
+    local repository_ok=true
+
+    # Check if snapshot repository already exists
+    if ! curl --fail -s u "elastic:$es_pass" -k "https://localhost:9200/_snapshot/beaker" > /dev/null ; then
```
Missing - before the `u` parameter.

Zalgo2462 reviewed (changes requested):
I've finished reading through the changes. Here's what I found on this first pass:
- I believe there is a bug where BeatsVersion is not passed to the agent install script when the script is started as a normal user. 
- The way we've been calling `curl` to interact with Elastic/ Kibana is insecure. There is a better way to run the command which will prevent leaking the credentials to normal users. 
- We need to ensure `python3` is available on the host system in `shell-lib` if we want to use a python script on the Docker host. Alternatively, we could find a way to run `check_kibana.py` in a Docker container. There may be other workarounds we could pursue such as using a tool like `yq` or `jq` in bash.

I also left a handful of questions, and a couple suggestions to add documentation/ comments.

caffeinatedpixel commented on docker-compose.yml:
```diff
@@ -1,9 +1,15 @@
+    driver_opts:
+      type: 'none'
+      device: /opt/BeaKer/snapshots
+      o: bind
```
If we ever needed to create a snapshot on a cluster, then the driver could be changed to a NFS, which is required for snapshotting clusters.

caffeinatedpixel commented on installer/stage/BeaKer/install_beaker.sh:
```diff
@@ -160,6 +236,352 @@ ensure_certificates_exist () {
+                        echo ""
+                        echo "The Elastic stack $(tput bold)cannot$(tput sgr0) be downgraded to a previous version."
+                        echo "In the event of upgrade failure, BeaKer must be reinstalled with the previously installed version."
+                        echo "The BeaKer installer will prompt you to create a snapshot that can be used to restore your data in the event of upgrade failure."
```
Delete any winlogbeat data streams and sysmon indices. Go to Stack Management > Snapshot & Restore in Kibana and restore the snapshot, matching on the index patterns of `winlogbeat-*, sysmon-*`.

caffeinatedpixel commented on installer/stage/BeaKer/install_beaker.sh:
```diff
@@ -96,9 +119,47 @@ ensure_env_file_exists () {
+    else
+        UPGRADE_INSTALL=true
+        
+        if ! $("$docker_sudo" grep -q "^KIBANA_SERVICE_TOKEN" "$BEAKER_CONFIG_DIR/env" ); then
```
This branch isn't solely for 7.5 -> 7.17. `UPGRADE_INSTALL` should be set to true for any upgrade install. Adding the KIBANA_SERVICE_TOKEN technically should only happen during an upgrade to 7.17, but if those fields are missing from the env file, Kibana will not start. The only danger I see from not checking CURRENT_ELASTIC_VERSION is if a user manually deleted some of the lines from the env file, which then that section would get appended to the env file. The user would need to delete any duplicate lines and validate the env file. I don't think it's necessarily worth it to diff the env file in this case.

Zalgo2462 commented:
I'm receiving the following error when trying to perform the initial upgrade from 7.5.2 to 7.17.9:
![image](https://user-images.githubusercontent.com/761220/231917033-8775db12-88c0-4a59-9d6c-acb1ca8f33b4.png)

Zalgo2462 commented on installer/generate_installer.sh:
```diff
@@ -93,7 +101,11 @@ if [ ! "$NO_BUILD" ]; then
-    $SUDO docker-compose build --pull $NO_CACHE $DOCKER_BUILD_SERVICES
+    for version in "${elk_versions[@]}"; do
+      v=$(echo $version|tr -d '\n')
+      #export ELK_STACK_VERSION="$v"
```
Please uncomment this line. It appears that commenting this line prevents docker-compose from labeling the resulting elastic and kibana images with the appropriate version tags.

Consider using the docker-compose.yml syntax for specifying the build-args for each of the containers individually instead of passing the build arg as a flag on the next line. 

Example which passes the ELK_STACK_VERSION environment variable in as a build argument:
```
build:
      context: ./elasticsearch
      args:
        ELK_STACK_VERSION: ${ELK_STACK_VERSION} 
```

caffeinatedpixel replied:
Uncommented the line. I did not convert to docker-compose syntax because it doesn't seem to tag the images as expected. I don't have the bandwidth to look into it at the moment, especially since the current tagging implementation isn't broken.

Zalgo2462 commented:
My testing environment consists of Zeek watching a Windows domain and sending that data to RITA/ AC-Hunter. As a part of that domain, I have a workstation running a C2 implant alongside the BeaKer agent. I ensured that the workstation was correctly reporting the C2 connection to the existing version of BeaKer (7.5.2). 

After working through the bug above, I performed an update to 7.17. I was able to confirm that the new version of the dashboard pulled in the old daily indexes from 7.5.2. Additionally, I was able to confirm that the old daily indexes continued to receive new events from the 7.5.2 winlogbeat agents. 

Next, I will upgrade the winlogbeat agent on the workstation running the implant to 7.17.9.

Zalgo2462 commented:
Would you be open to adding a blurb to the end of the upgrade install depending on the resulting version? A reminder for the user to go upgrade their winlogbeat agents to the appropriate version and then to come back and run the next upgrade (if needed) might be helpful.

Zalgo2462 commented:
I was able to successfully upgrade winlogbeats to 7.19 and I was able to confirm that it was sending data to the 7.19 data stream in elastic. The upgrade to 7.19 went well all around. I was able to confirm the ILM policies, data streams, index template, and dashboards were all properly upgraded or created.  

During the upgrade to 8.x I received the following error:

![image](https://user-images.githubusercontent.com/761220/231929051-49cbef1c-c155-490d-89e8-079a47bf3a4c.png)

Zalgo2462 commented:
It appears that the elasticsearch group does not have write access to the snapshots folder in the container causing the error above:
![image](https://user-images.githubusercontent.com/761220/231933793-1848df96-9b1d-48fe-9718-77a019b68b0d.png)

Please confirm this is the cause of the error and add a `chmod` to the Dockerfile to remedy the issue if needed.

caffeinatedpixel commented:
> Please confirm this is the cause of the error and add a `chmod` to the Dockerfile to remedy the issue if needed.

Added chmod to give write perms, which matches the permissions of many other folders within the elasticsearch container. 
Also added a check to start BeaKer if it isn't already running before creating a snapshot.

Zalgo2462 commented:
I was able to build a new installer with the latest changes. I copied it over to my system running 7.17 and selected the option to reinstall. 

This took care of the snapshot issue, and I was able to re-run the installer to upgrade to 8.x. 

The post-installation messages look good to me. 

I'll test the final upgrade for the winlogbeat agents tonight.

Zalgo2462 reviewed (approved):
When running the powershell install script to upgrade the winlogbeat agent to 8.x, I saw the following error, but I couldn't tell what had actually gone wrong. Everything seemed to still work. I tried to reproduce the error, but I couldn't get it to pop up again.
![image](https://user-images.githubusercontent.com/761220/232653817-0998b876-57e6-4bc8-bc8e-961c715d289d.png)

The 8.x agent appears to work well.
